### ISSUE TITLE
Expose memory allocation to user applications

### ISSUE THREAD

**Issue by Thalhammer** [libuv/libuv#3262]
* **Version**: Any
* **Platform**: Any

Currently there is no official way to allocate memory using the allocator set for libuv or get said allocator.
My proposal is to make several small, non abi/api breaking changes to allow for this.
1. Make `struct uv__allocator_t` public and add function `void uv_get_allocator(struct uv__allocator_t* info);` to get a copy of the internal struct. An alternative would be to provide 4 pointers to function pointers.
2. Add `uv_malloc`,`uv_realloc`,`uv_calloc` and `uv_free` calling the relevant internal functions (same name, but double underscore).

The background for this is that when writing libraries that interact with libuv you don't necessarily know if the user set a custom allocator for libuv. While the library could allocate all structures it needs using the generic malloc/free and be fine, the user probably had a good reason to set a custom allocator and it should probably be respected. In addition to this it would allow freeing memory allocated and returned by libuv inside the application without worrying about dynamic library boundaries. Given the small changes required I think its a worthwhile change.

I am aware that I can just `extern "C"` the internal functions and it in all my cases where libuv is statically linked its fine, however it might not work if libuv is used as a shared library and the signature of those functions could change at any time (they wont given its malloc, but they could since its not part of the abi promise).

I am willing to implement it and do a PR if theres hope it gets merged.

**Comment by vtjnash** [comment]
He wants a way to get that list of functions back out, after setting them with that API

**Comment by Thalhammer** [comment]
> Don't get me wrong, but it looks like you want to re-implement

No I want to implement a set of functions adding to uv_replace_allocator that allows a user to allocate memory using the memory functions used by libuv (malloc by default or something else if the user changed them).

TLDR: I want to make uv__malloc and uv__free user facing.

**Comment by bnoordhuis** [comment]
I think most libuv consumers that override the memory allocator do so in order to track libuv's internal memory usage.

Such users won't expect third-party memory allocations to show up in their graphs. I'm inclined to leave well enough alone because we try very hard to not break existing users.

**Comment by Thalhammer** [comment]
Closing due to lack of interest.
